File: python/src/commands.py
import re
from .farm import Farm

class Commands:
    """
    Class that works as a kind of UI

    ...
    Attributes
    ----------
    app : Farm
        The app that will handle the actions
    data : list
        Input array taken from sys.argv
    data_len : int
        Input array length

    Methods
    -------
    process()
        Obtain the command and process the input data

    get_command()
        Return the command

    __verify_command()
        Identify the acction to execute

    __list_command()
        Process the [list | l] command

    __create_command()
        Process the [create | c] command

    __delete_command()
        Process the [delete | c] command

    __search_command()
        Process the [search | s] command

    __food_command()
        Process the [food   | f] command

    __wash_command()
        Process the [wash   | w] command

    __alive_command()
        Process the [alive  | a] command

    __get_param(value, data_type)
        Returns the queried param

    __create_command_help()
        Shows help for the [create | c] command

    __delete_command_help()
        Shows help for the [delete | c] command

    __search_command_help()
        Shows help for the [search | s] command

    __food_command_help()
        Shows help for the [food   | f] command

    __wash_command_help()
        Shows help for the [wash   | w] command

    __alive_command_help()
        Shows help for the [alive  | a] command

    __help()
        Shows help for the app
    """
    def __init__(self, data):
        """
        Parameters
        ----------
        data : list
            The input array
        """
        self.app = Farm()
        self.data = data
        self.data_len = len(self.data)

    def process(self):
        """
        Obtain the command and process the input data

        Return
        ------
        None
        """
        command = self.get_command()

        if command:
            self.__verify_command(command)
        else:
            self.__help()

    def get_command(self):
        """
        Returns the command

        Return
        ------
        command : str
            The second item in the input array, if it's given
        None
            otherwise
        """
        command = None

        if self.data_len > 1:
            command = self.data[1]

        return command

    def __verify_command(self, command):
        """
        Identify the specific action to execute and call the respective action

        ...
        Parameters
        ----------
        command : str
            Command obtained from get_command() method

        Return
        ------
        None
        """
        if command in ["list", "l"]:
            self.__list_command()
        elif command in ["create", "c"]:
            if self.__create_command():
                self.__list_command()   # Showing animals
        elif command in ["delete", "d"]:
            if self.__delete_command():
                self.__list_command()   # Showing animals
        elif command in ["search", "s"]:
            self.__search_command()
        elif command in ["food", "f"]:
            self.__food_command()
        elif command in ["wash", "w"]:
            self.__wash_command()
        elif command in ["alive", "a"]:
            self.__alive_command()
        elif command in ["h", "help"]:
            self.__help()
        else:
            self.__help()

    def __list_command(self):
        """
        Process the [list | l] command

        Return
        ------
        None
        """
        age = self.__get_param("age=", "(\d+)")
        animal_type = self.__get_param("type=", "([a-zA-Z]+)")

        self.app.list_animals(age=age, animal_type=animal_type)

    def __create_command(self):
        """
        Process the [create | c] command

        Return
        ------
        True
            if the animal was created
        """
        name = self.__get_param("name=", "([a-zA-Z]+)")
        animal_type = self.__get_param("type=", "([a-zA-Z]+)")

        created = False

        if name:
            created = self.app.create_animal(name, animal_type)
        else:
            print("Error!: A name is required to create an animal")
            self.__create_command_help()

        return created

    def __delete_command(self):
        """
        Process the [delete | c] command

        Return
        ------
        True
            if the animal was deleted
        """
        name = self.__get_param("name=", "([a-zA-Z]+)")

        deleted = False
        if name:
            deleted = self.app.delete_animal(name)
        else:
            print("Error!: A name is required to delete a animal")
            self.__delete_command_help()
        return deleted

    def __search_command(self):
        """
        Process the [search | s] command

        Return
        ------
        None
        """
        name = self.__get_param("name=", "([a-zA-Z]+)")
        if name:
            self.app.search(name)
        else:
            print("Error!: A name is required to search a animal")
            self.__search_command_help()

    def __food_command(self):
        """
        Process the [food   | f] command

        Return
        ------
        None
        """
        name = self.__get_param("name=", "([a-zA-Z]+)")
        if name:
            self.app.food(name)
        else:
            print("Error!: A name is required to give food to a animal")
            self.__food_command_help()

    def __wash_command(self):
        """
        Process the [wash   | w] command

        Return
        ------
        None
        """
        name = self.__get_param("name=", "([a-zA-Z]+)")
        if name:
            self.app.wash(name)
        else:
            print("Error!: A name is required to give a shower to a animal")
            self.__wash_command_help()

    def __alive_command(self):
        """
        Process the [alive  | a] command

        Return
        ------
        None
        """
        name = self.__get_param("name=", "([a-zA-Z]+)")
        if name: 
            self.app.alive(name)
        else:
            print("Error!: A name is required to show if a animal is alive")
            self.__alive_command_help()

    def __get_param(self, value, data_type):
        """
        Returns the queried param using regex

        Parameters
        ----------
        value : str
            variable name to search
        data_type : str
            variable pattern equivalent

        Return
        ------
        param : str
            the param value if a match occurs
        None
            otherwise
        """
        param = None

        if self.data_len > 2:
            data_str = ';'.join(self.data[2:])
            param = re.search(value+data_type, data_str)
            if param:
                param = param.group(1).replace(value, '') if param.group(1) else None

        return param

    def __create_command_help(self):
        """
        Shows help for the [create | c] command

        Return
        ------
        None
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Command Usage:")
        print("  [create  | c]   <name=> <type=>    creates a animal with name")
        print("")
        print("Examples:")
        print("  python main.py [create  | c]  name=my_animal_name")
        print("  python main.py [create  | c]  name=my_animal_name type=pig")

    def __delete_command_help(self):
        """
        Shows help for the [delete | c] command

        Return
        ------
        None
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Command Usage:")
        print("  [delete  | d]   <name=>        deletes a animal by name")
        print("")
        print("Examples:")
        print("  python main.py [delete  | d]  name=my_animal_name")

    def __search_command_help(self):
        """
        Shows help for the [search | s] command

        Return
        ------
        None
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Command Usage:")
        print("  [search  | s]   <name=>       searchs a animal by name")
        print("")
        print("Examples:")
        print("  python main.py [search  | s]  name=my_animal_name")

    def __food_command_help(self):
        """
        Shows help for the [food   | f] command

        Return
        ------
        None
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Command Usage:")
        print("  [food  | f]   <name=>       gives food to a animal by name")
        print("")
        print("Examples:")
        print("  python main.py [food  | f]  name=my_animal_name")

    def __wash_command_help(self):
        """
        Shows help for the [wash   | w] command

        Return
        ------
        None
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Command Usage:")
        print("  [wash  | w]   <name=>       gives a shower to a animal by name")
        print("")
        print("Examples:")
        print("  python main.py [wash  | w]  name=my_animal_name")

    def __alive_command_help(self):
        """
        Shows help for the [alive  | a] command

        Return
        ------
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Command Usage:")
        print("  [alive  | a]   <name=>       checks if an animal is alive by name")
        print("")
        print("Examples:")
        print("  python main.py [alive  | a]  name=my_animal_name")

    def __help(self):
        """
        Shows help for the app

        Return
        ------
        None
        """
        print("The Python Farm - a simple command line animals app")
        print()
        print("Usage:")
        print("  python main.py [command]")
        print()
        print("Available Commands:")
        print("  [list    | l]   <age=>  <type=>    list all available animals")
        print("  [create  | c]   <name=> <type=>    create a animal with name")
        print("  [delete  | d]   <name=>            delete a animal")
        print("  [search  | s]   <name=>            search a animal")
        print("  [food    | f]   <name=>            give food to a animal")
        print("  [wash    | w]   <name=>            give a shower to a animal")
        print("  [alive   | a]   <name=>            show if a animal is alive")
        print("  [help    | h]                      help about commands")
        print("")
        print("Flags:")
        print("  -v, --version  show the app version")
        print("")
        print("Use python main.py [command] --help for more information about a command.")

File: golang/cmd/src/cli/commands.go
// Package cli provides to the user a way to interact with the app.
// It uses os.Args to receive the user input an process it.
package cli

import (
	"fmt"
	"regexp"
	"strings"
)

// Commands works as a kind of UI
// It process the user input and route it to the respective Farm action
type Commands struct {
	App  Farm     // The app that will handle the actions
	Data []string // Input array taken from os.Args
}

// NewCommands creates a new Commands instance,
// and setups the specific Farm properties.
//
// It assigns the input data to work with it.
func NewCommands(data []string) Commands {
	commands := Commands{}
	commands.App = NewFarm()
	commands.Data = data

	return commands
}

// Process Obtains the command and process the input data.
// If no command is given it show the Help, otherwise
// it uses the VerifyCommand to route the actions
func (c *Commands) Process() {
	command := c.GetCommand()

	if command != nil {
		c.VerifyCommand(*command)
	} else {
		c.Help()
	}
}

// GetCommand returns the main command,
// That tells what action will be performed
//
// It returns the main command, or nil if not command was given
func (c *Commands) GetCommand() *string {
	if len(c.Data) > 1 {
		return &(c.Data[1])
	} else {
		return nil
	}
}

// VerifyCommand takes the main command, identifies the specific action to execute
// and call the respective action.
//
// It uses the Contains custom method, to identify what action to execute. Also,
// it give the names to the commands.
//
// The param command is obtained from the GetCommand method
func (c *Commands) VerifyCommand(command string) {
	if c.Contains([]string{"list", "l"}, command) {
		c.ListCommand()
	} else if c.Contains([]string{"create", "c"}, command) {
		if c.CreateCommand() {
			c.ListCommand()
		}
	} else if c.Contains([]string{"delete", "d"}, command) {
		if c.DeleteCommand() {
			c.ListCommand()
		}
	} else if c.Contains([]string{"search", "s"}, command) {
		c.SearchCommand()
	} else if c.Contains([]string{"food", "f"}, command) {
		c.FoodCommand()
	} else if c.Contains([]string{"wash", "w"}, command) {
		c.WashCommand()
	} else if c.Contains([]string{"alive", "a"}, command) {
		c.AliveCommand()
	} else if c.Contains([]string{"help", "h"}, command) {
		c.Help()
	} else {
		c.Help()
	}
}

// ListCommand process the [list | l] command
func (c *Commands) ListCommand() {
	c.App.ListAnimals()
}

// CreateCommand process the [create | c] command
// It uses the GetParam methods to identify the user input
// and call the respective Farm action
//
// It return true if the animal was created, false otherwise
func (c *Commands) CreateCommand() bool {
	created := false
	name := c.GetParam("name=", "([a-zA-Z]+)")
	animalType := c.GetParam("type=", "([a-zA-Z]+)")

	if name != nil {
		if animalType == nil {
			// An auxiliar variable is created because
			// the GetParam method returns nil if doesn't find a match
			// and we cannot assign a string o a nil pointer
			defaultType := "cow"
			animalType = &defaultType
		}

		created = c.App.CreateAnimal(*name, *animalType)
	} else {
		fmt.Println("Error!: A name is required to create an animal")
		c.CreateCommandHelp()
	}

	return created
}

// DeleteCommand process the [delete | c] command
// It uses the GetParam methods to identify the user input
// and call the respective Farm action
//
// It return true if the animal was deleted, false otherwise
func (c *Commands) DeleteCommand() bool {
	deleted := false
	name := c.GetParam("name=", "([a-zA-Z]+)")

	if name != nil {
		deleted = c.App.DeleteAnimal(*name)
	} else {
		fmt.Println("Error!: A name is required to delete a animal")
		c.DeleteCommandHelp()
	}

	return deleted
}

// SearchCommand process the [search | s] command
// It uses the GetParam methods to identify the user input
// and call the respective Farm action
func (c *Commands) SearchCommand() {
	name := c.GetParam("name=", "([a-zA-Z]+)")

	if name != nil {
		c.App.Search(*name)
	} else {
		fmt.Println("Error!: A name is required to search a animal")
		c.SearchCommandHelp()
	}
}

// FoodCommand process the [food | f] command
// It uses the GetParam methods to identify the user input
// and call the respective Farm action
func (c *Commands) FoodCommand() {
	name := c.GetParam("name=", "([a-zA-Z]+)")

	if name != nil {
		c.App.Food(*name)
	} else {
		fmt.Println("Error!: A name is required to give food to a animal")
		c.FoodCommandHelp()
	}
}

// WashCommand process the [wash | w] command
// It uses the GetParam methods to identify the user input
// and call the respective Farm action
func (c *Commands) WashCommand() {
	name := c.GetParam("name=", "([a-zA-Z]+)")

	if name != nil {
		c.App.Wash(*name)
	} else {
		fmt.Println("Error!: A name is required to give a shower to a animal")
		c.WashCommandHelp()
	}
}

// AliveCommand process the [alive | a] command
// It uses the GetParam methods to identify the user input
// and call the respective Farm action
func (c *Commands) AliveCommand() {
	name := c.GetParam("name=", "([a-zA-Z]+)")

	if name != nil {
		c.App.Alive(*name)
	} else {
		fmt.Println("Error!: A name is required to show if a animal is alive")
		c.AliveCommandHelp()
	}
}

// GetParam returns the queried param using regex patterns
// It joins the input data into a string and applies a regex pattern
//
// The param value is the variable name to search
// The param dataType is the variable pattern equivalent
//
// It returns a pointer of the string if a match was found, nil otherwise
//   Examples:
//	    data := []string{"name=jay", "age=10"}
//
//		value := GetParam("name=", "([a-zA-Z]+)")
//		fmt.Println(*value)
//		  => "jay"
//
//		value := GetParam("age=", "(\\d+)")
//		fmt.Println(*value)
//		  => 10
//
//		value := GetParam("type=", "([a-zA-Z]+)")
//		fmt.Println(*value)
//		  => nil
func (c *Commands) GetParam(value string, dataType string) *string {
	data := strings.Join(c.Data[1:], ";")

	r, _ := regexp.Compile(value + dataType)
	result := r.FindString(data)

	if result == "" {
		return nil
	}

	result = strings.ReplaceAll(result, value, "")

	if result == "" {
		return nil
	}

	return &result
}

// CreateCommandHelp shows help for the [create | c] command
func (c *Commands) CreateCommandHelp() {
	fmt.Println("The Go Farm - a simple command line animals app")
	fmt.Println()
	fmt.Println("Command Usage:")
	fmt.Println("  [create  | c]   <name=> <type=>    creates a animal with name")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  ./farm [create  | c] name=my_animal_name")
	fmt.Println("  ./farm [create  | c] name=my_animal_name type=pig")
}

// DeleteCommandHelp shows help for the [delete | c] command
func (c *Commands) DeleteCommandHelp() {
	fmt.Println("The Go Farm - a simple command line animals app")
	fmt.Println()
	fmt.Println("Command Usage:")
	fmt.Println("  [delete  | d]   <name=>        deletes a animal by name")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  ./farm [delete  | d] name=my_animal_name")
}

// SearchCommandHelp shows help for the [search | s] command
func (c *Commands) SearchCommandHelp() {
	fmt.Println("The Go Farm - a simple command line animals app")
	fmt.Println()
	fmt.Println("Command Usage:")
	fmt.Println("  [search  | s]   <name=>       searchs a animal by name")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  ./farm [search  | s] name=my_animal_name")
}

// FoodCommandHelp shows help for the [food | f] command
func (c *Commands) FoodCommandHelp() {
	fmt.Println("The Go Farm - a simple command line animals app")
	fmt.Println()
	fmt.Println("Command Usage:")
	fmt.Println("  [food  | f]   <name=>       gives food to a animal by name")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  ./farm [food  | f] name=my_animal_name")
}

// WashCommandHelp shows help for the [wash | w] command
func (c *Commands) WashCommandHelp() {
	fmt.Println("The Go Farm - a simple command line animals app")
	fmt.Println()
	fmt.Println("Command Usage:")
	fmt.Println("  [wash  | w]   <name=>       gives a shower to a animal by name")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  ./farm [wash  | w] name=my_animal_name")
}

// AliveCommandHelp shows help for the [alive | a] command
func (c *Commands) AliveCommandHelp() {
	fmt.Println("The Go Farm - a simple command line animals app")
	fmt.Println()
	fmt.Println("Command Usage:")
	fmt.Println("  [alive  | a]   <name=>       checks if an animal is alive by name")
	fmt.Println("")
	fmt.Println("Examples:")
	fmt.Println("  ./farm [alive  | a] name=my_animal_name")
}

// Help shows help for the app
func (c *Commands) Help() {
	fmt.Println("The Golang Farm - a simple command line animals app")
	fmt.Println("")
	fmt.Println("Usage:")
	fmt.Println("  ./farm [command]")
	fmt.Println("")
	fmt.Println("Available Commands:")
	fmt.Println("  [list    | l]                      list all available animals")
	fmt.Println("  [create  | c]   <name=> <type=>    create a animal with name")
	fmt.Println("  [delete  | d]   <name=>            delete a animal")
	fmt.Println("  [search  | s]   <name=>            search a animal")
	fmt.Println("  [food    | f]   <name=>            give food to a animal")
	fmt.Println("  [wash    | w]   <name=>            give a shower to a animal")
	fmt.Println("  [alive   | a]   <name=>            show if a animal is alive")
	fmt.Println("  [help    | h]                      help about commands")
	fmt.Println("")
	fmt.Println("Flags:")
	fmt.Println("Use ./farm help for more information about a command.")
}

// Contains determine if an array of string contains a specific element.
//
// The param array is the array of string to check if contains an element
// The param value is the element to search
//
// It returns true if the array contains the element, false otherwise
func (c *Commands) Contains(array []string, value string) bool {
	found := false
	for _, val := range array {
		if val == value {
			found = true
			break
		}
	}

	return found
}
